src/utils/charts.js

```diff
@@ -0,0 +1,255 @@
+      }-${entryDate.getFullYear()}`,
+      count,
+    };
+  });
+};
+
+// given an actionsByDay object, finds max value to set on the yAxis in the graph in ActionsByDayChart.js
+export const findYAxisMax = (actionsByDay) => {
+  const arrayOfActionsCount = Object.values(actionsByDay);
+  if (!arrayOfActionsCount.length) {
+    return null;
+  }
+  const maxActionsCount = arrayOfActionsCount.reduce((a, b) => Math.max(a, b));
+  let yAxisMax;
+  // if maxActionsCount <= 100, round up yAxisMax to the nearest ten; else, to the nearest hundred
+  if (maxActionsCount <= 100) {
+    yAxisMax = Math.ceil(maxActionsCount / 10) * 10;
+  } else {
+    yAxisMax = Math.ceil(maxActionsCount / 100) * 100;
+  }
+  return yAxisMax;
+};
+
+// helper function used in getActionsByTimeOfDay
+// todo: update this function to retrieve hour of day using JS Date objects/moment
+const getActionHourOfDay = (action) => {
+  // expects action to be an object with a createdAt key
+  // createdAt should have the format "2020-12-31T23:59:59.999Z"
+  return action.createdAt.slice(11, 13);
+};
+
+// Takes array of action objects and returns an object with {key: value} pairs of {hourOfDay: #-of-actions}
+export const getActionsByTimeOfDay = (actions) => {
+  const actionsByTimeOfDay = {
+    [LATE_NIGHT]: 0,
+    [EARLY_MORNING]: 0,
+    [MORNING]: 0,
+    [AFTERNOON]: 0,
+    [EVENING]: 0,
+    [NIGHT]: 0,
+  };
+  actions.forEach((action) => {
+    const actionHourOfDay = getActionHourOfDay(action);
+    if (actionHourOfDay >= 0 && actionHourOfDay < 4) {
+      actionsByTimeOfDay[LATE_NIGHT] += 1;
+    } else if (actionHourOfDay >= 4 && actionHourOfDay < 8) {
+      actionsByTimeOfDay[EARLY_MORNING] += 1;
+    } else if (actionHourOfDay >= 8 && actionHourOfDay < 12) {
+      actionsByTimeOfDay[MORNING] += 1;
+    } else if (actionHourOfDay >= 12 && actionHourOfDay < 16) {
+      actionsByTimeOfDay[AFTERNOON] += 1;
+    } else if (actionHourOfDay >= 16 && actionHourOfDay < 20) {
+      actionsByTimeOfDay[EVENING] += 1;
+    } else if (actionHourOfDay >= 20 && actionHourOfDay < 24) {
+      actionsByTimeOfDay[NIGHT] += 1;
+    } else {
+      // eslint-disable-next-line no-console
+      console.log(`actionHourOfDay ${actionHourOfDay} is undefined`);
+    }
+  });
+  return actionsByTimeOfDay;
+};
+
+// Takes object with {key: value} pairs of {timeOfDay: #-of-actions}
+// returns a date-sorted array in Recharts.js format
+export const formatActionsByTimeOfDay = (actionsByTimeOfDayObject) => {
+  const actionsByTimeOfDayArray = Object.entries(actionsByTimeOfDayObject);
+  return actionsByTimeOfDayArray.map(([timeOfDay, count]) => {
+    return {
+      timeOfDay,
+      count,
+    };
+  });
+};
+
+// Takes array of action objects and returns an object with {key: value} pairs of {verb: %-of-actions}
+export const getActionsByVerb = (actions) => {
+  const totalActionsCount = actions.length;
+  const actionsByVerbCount = {};
+  actions.forEach((action) => {
```

Comment:
You don't have to update this, but this could have been done using either:
- groupby (lodash) on the actions verb would give you a map whose keys would be the actions and the values would be an array of all corresponding actions. Then a map can be used to have the length of the array.
- reduce: given an accumulator `{ }` you can go through `actions` and add them to the map "one by one". It looks like a lot to your foreach procedure but the accumulator is a bit more transparent.

I prefer the first option as it is much more clearer.

Reply:
both interesting!